### ISSUE TITLE
XrdHttp fails to refresh CRLs

### ISSUE THREAD

**Issue by olifre** [xrootd/xrootd#750]
Using xrootd 4.8.3, I observe the following after a few days:
```
139990299645696:error:80066405:lib(128):func(102):reason(1029):sslutils.c:1915:
139990299645696:error:80066411:lib(128):func(102):reason(1041):sslutils.c:2110:: CRL has expired [subject=/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=ddmadmin/CN=531497/CN=Robot: ATLAS Data Management,issuer=/DC=ch/DC=cern/CN=CERN Grid Certification Authority]
139990299645696:error:14089086:SSL routines:ssl3_get_client_certificate:certificate verify failed:s3_srvr.c:3327:
```
all over in our logs. 

Checking:
```
# ls -lart b4278411.r0 c2a48ab6.r0 5168735f.r0 4339b4bc.r0
-rw-r--r--. 1 root root  1535 19. Jun 12:47 c2a48ab6.r0
-rw-r--r--. 1 root root  1535 19. Jun 12:47 b4278411.r0
-rw-r--r--. 1 root root 33292 19. Jun 12:47 5168735f.r0
-rw-r--r--. 1 root root 33292 19. Jun 12:47 4339b4bc.r0
```
shows they are up-to-date (those are the CRLs for CERN-GridCA and CERN-Root-2, we run fetchcrl regulartly). 

Checking xrootd with `strace`, I find:
```
[pid 13040] futex(0x610f18, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
[pid 10153] <... accept4 resumed> {sa_family=AF_INET6, sin6_port=htons(51392), inet_pton(AF_INET6, "::ffff:128.142.132.207", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, [28], SOCK_CLOEXEC) = 25
[pid 10153] setsockopt(25, SOL_SOCKET, SO_LINGER, {onoff=1, linger=3}, 8) = 0
[pid 10153] setsockopt(25, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0
[pid 10153] setsockopt(25, SOL_TCP, TCP_NODELAY, [1], 4) = 0
[pid 10153] futex(0x7ffc4e2a27b0, FUTEX_WAKE_PRIVATE, 1) = 1
[pid  3022] <... futex resumed> )       = 0
[pid 10153] poll([{fd=25, events=POLLIN|POLLRDNORM}], 1, 30000 <unfinished ...>
[pid  3022] futex(0x610f18, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 13039] <... futex resumed> )       = 0
[pid 10153] <... poll resumed> )        = 1 ([{fd=25, revents=POLLIN|POLLRDNORM}])
[pid  3022] <... futex resumed> )       = 1
[pid 13039] futex(0x610f38, FUTEX_WAIT_PRIVATE, 2, NULL <unfinished ...>
[pid 10153] recvfrom(25,  <unfinished ...>
[pid  3022] futex(0x610f38, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 13039] <... futex resumed> )       = -1 EAGAIN (Resource temporarily unavailable)
[pid 10153] <... recvfrom resumed> "\26\3\1\1\24\1\0\1\20\3\3[)'\305\261\351\17<\225C\232\311\27V\240\221q\t\377\26\275"..., 44, MSG_PEEK, NULL, NULL) = 44
[pid  3022] <... futex resumed> )       = 0
[pid 13039] futex(0x610f38, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 10153] poll([{fd=25, events=POLLIN|POLLRDNORM}], 1, 10000 <unfinished ...>
[pid  3022] futex(0x7ffc4e2a27b0, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
[pid 13039] <... futex resumed> )       = 0
[pid 10153] <... poll resumed> )        = 1 ([{fd=25, revents=POLLIN|POLLRDNORM}])
[pid 13039] accept4(17,  <unfinished ...>
[pid 10153] recvfrom(25, "\26\3\1\1\24\1\0\1\20\3\3[)'\305\261", 16, MSG_PEEK, NULL, NULL) = 16
[pid 10153] epoll_ctl(7, EPOLL_CTL_ADD, 25, {0, {u32=738201208, u64=139990902247032}}) = 0
[pid 10153] setsockopt(25, SOL_SOCKET, SO_RCVTIMEO, "\n\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
[pid 10153] setsockopt(25, SOL_SOCKET, SO_SNDTIMEO, "\n\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 16) = 0
[pid 10153] read(25, "\26\3\1\1\24", 5) = 5
[pid 10153] read(25, "\1\0\1\20\3\3[)'\305\261\351\17<\225C\232\311\27V\240\221q\t\377\26\275\232G\260\300\276"..., 276) = 276
[pid 10153] write(25, "\26\3\3\0:\2\0\0006\3\3\24\275\332\341L+Tw:\2216\231\1\352~\213\37S\v\274."..., 2216) = 2216
[pid 10153] read(25, "\26\3\3\0378", 5) = 5
[pid 10153] read(25, "\v\0\0374\0\0371\0\3L0\202\3H0\202\2\261\240\3\2\1\2\2\1\0000\r\6\t*\206"..., 7992) = 7992
[pid 10153] stat("/etc/grid-security/certificates/c2a48ab6.r1", 0x7f523588a1f0) = -1 ENOENT (No such file or directory)
[pid 10153] stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=2309, ...}) = 0
[pid 10153] stat("/etc/grid-security/certificates/c2a48ab6.r1", 0x7f523588a1f0) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/e187c0c8.signing_policy", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/e187c0c8.namespaces", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/f27afd4e.signing_policy", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/f27afd4e.namespaces", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/4ad425d0.signing_policy", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/4ad425d0.namespaces", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/7768cc9b.signing_policy", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/7768cc9b.namespaces", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/5168735f.signing_policy", O_RDONLY) = 26
[pid 10153] ioctl(26, TCGETS, 0x7f5235889900) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 10153] read(26, "# @(#)$Id: 4339b4bc.signing_poli"..., 8192) = 269
[pid 10153] read(26, "", 8192)          = 0
[pid 10153] ioctl(26, TCGETS, 0x7f5235889910) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 10153] close(26)                   = 0
[pid 10153] open("/etc/grid-security/certificates/5168735f.namespaces", O_RDONLY) = -1 ENOENT (No such file or directory)
[pid 10153] open("/etc/grid-security/certificates/c2a48ab6.signing_policy", O_RDONLY) = 26
[pid 10153] ioctl(26, TCGETS, 0x7f5235889900) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 10153] read(26, "# @(#)$Id: b4278411.signing_poli"..., 8192) = 362
[pid 10153] read(26, "", 8192)          = 0
[pid 10153] ioctl(26, TCGETS, 0x7f5235889910) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 10153] close(26)                   = 0
[pid 10153] open("/etc/grid-security/certificates/c2a48ab6.namespaces", O_RDONLY) = 26
[pid 10153] ioctl(26, TCGETS, 0x7f5235889900) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 10153] read(26, "################################"..., 8192) = 774
[pid 10153] read(26, "", 8192)          = 0
[pid 10153] ioctl(26, TCGETS, 0x7f5235889910) = -1 ENOTTY (Inappropriate ioctl for device)
[pid 10153] close(26)                   = 0
[pid 10153] stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=2309, ...}) = 0
[pid 10153] stat("/etc/grid-security/certificates/5168735f.r1", 0x7f523588a1f0) = -1 ENOENT (No such file or directory)
[pid 10153] write(25, "\25\3\3\0\2\2-", 7) = 7
[pid 10153] write(2, "139991062198016:error:80066405:l"..., 80) = 80
[pid 10153] write(2, "139991062198016:error:80066411:l"..., 261) = 261
[pid 10153] write(2, "139991062198016:error:14089086:S"..., 114) = 114
```

Any ideas what could be causing that? 

Restarting the xrootd service immediately gets rid of that issue.

**Comment by bbockelm** [comment]
Hi @olifre -

Apologies, I somehow overlooked the creation of this ticket!

At Nebraska, we have not observed this behavior.  Can you please post your config file so I can see the difference in the setup?

For your original error message - what log does that come from?

Brian

**Comment by olifre** [comment]
Hi @bbockelm ,

I'll attach the config I used in June below (hooray for git-managed configuration management ;-) ). We have only tuned unrelated config parameters since then, but since we perform daily service restarts, the issue is completely gone for now. 

> For your original error message - what log does that come from?

That's straight from the `xrootd.log`. 

Cheers and thanks for looking into it,
Oliver

----------------------------------
```
acc.authdb /etc/xrootd/auth_file-grid
acc.authrefresh 60
all.export /cephfs/grid/atlas/atlaslocalgroupdisk r/w
all.export /cephfs/grid/atlas/atlasscratchdisk r/w
all.export /cephfs/grid/atlas/user/scratch r/w
all.export /cephfs/grid/ops r/w
all.manager xrootd.physik.uni-bonn.de:1213
all.role server
all.role manager if xrootd.physik.uni-bonn.de
all.sitename UNI-BONN
cms.allow host xrootd.physik.uni-bonn.de
cms.allow host xrootd*.physik.uni-bonn.de
cms.dfs limit 0 lookup distrib mdhold 0 redirect immed retries 2
http.cadir /etc/grid-security/certificates
http.cert /etc/grid-security/hostcert.pem
http.key /etc/grid-security/hostkey.pem
http.secretkey REDACTEDHERE
http.secxtractor /usr/lib64/libXrdHttpVOMS.so
if exec xrootd
xrd.protocol XrdHttp /usr/lib64/libXrdHttp.so
fi
if xrootd.physik.uni-bonn.de
else
http.selfhttps2http no
fi
if xrootd.physik.uni-bonn.de
http.desthttps yes
fi
ofs.authorize 
ofs.tpc autorm ttl 180 1800 pgm /usr/local/bin/xrdcp-voms --server
sec.protocol /usr/lib64 gsi -ca:1 -crl:3 -gridmap:/dev/null -cert:/etc/grid-security/hostcert.pem -key:/etc/grid-security/hostkey.pem -certdir:/etc/grid-security/certificates
sec.protparm gsi -vomsfun:/usr/lib64/libXrdSecgsiVOMS.so -vomsfunparms:certfmt=raw|vos=atlas,ops|grps=/atlas,/ops
xrd.port 1094
xrootd.chksum adler32 crc32 md5
xrootd.seclib /usr/lib64/libXrdSec.so
```

**Comment by bbockelm** [comment]
@olifre - it's possible that this failure is coming from `libXrdHttpVOMS.so`; that's the most significant setup difference between your and my configs.  Any chance it's possible to try without?

The next version of `libXrdLcmaps.so` can likely be a drop-in replacement.  Since that will be released in January (and you'll be getting a new test setup around then), can we put the ticket on hold until then?

**Comment by olifre** [comment]
@bbockelm Sadly, testing without `libXrdHttpVOMS.so` in the production system is not really an option. But indeed, your feeling that this plugin may cause the issue sounds likely and would explain why most others who use Lcmaps have not seen it. 



> The next version of `libXrdLcmaps.so` can likely be a drop-in replacement. Since that will be released in January (and you'll be getting a new test setup around then), can we put the ticket on hold until then?

That's a very good idea! Yes, of course, no problem with that. Daily restarts work very well as a workaround for now. 
I dearly hope we can manage to get the test setup early next year, since we also have to commission a new building (with everything, networking, servers etc). 2019 wll be a fun year! :wink:

**Comment by simonmichal** [comment]
Hi guys, is this discussion still open, is there an action needed, or can we close it?

**Comment by olifre** [comment]
> The next version of `libXrdLcmaps.so` can likely be a drop-in replacement.

Dear Brian,
do you know if this applies to the new version, and if so, are there instructions on how to use that as drop-in replacement? Then I would test this out. 
Cheers and thanks,
Oliver

**Comment by bbockelm** [comment]
@olifre - your best bet is this documentation: https://opensciencegrid.org/docs/data/xrootd/xrootd-authorization/

**Comment by olifre** [comment]
@bbockelm That looks rather heavy - it seems to require the full LCMAPS setup, with unix accounts for each VO and a vomsmapfile. While the full setup may be used to replace `libXrdHttpVOMS` in some way, it shifts authorization away from the `auth_file` and seems to be too much functionality and necessary configuration for what we need. So it still does not seem to be a drop-in replacement to me, but rather an alternative approach (which comes with benefits of course if you also have a CE).

**Comment by bbockelm** [comment]
Ah - if you don't need all that, then you can set:

`-authzfunparms:no-authz`

And that will skip the `lcmaps` pieces and leave all the authorization to Xrootd.

**Comment by ffurano** [comment]
Hi,
 actually my test setup used to stay up for months without restarts, but... there were no CRLs.
 XrdHttpVoms does its own sec business through openssl and libvoms, so it may be that it never reloads
the CRLs, as the init phase is executed only when the plugin is loaded. 

I think that the daemon restarts are an effective workaround, which is what e.g. Apache does for the same case. Another workaround is making sure that there are no CRLs around, don't know if this is acceptable.

 A more elegant attempt could be to internally reinitialize it every few hours by invoking destroy_SSL_proxy_handler/setup_SSL_proxy_handler, I'll give it a try.

**Comment by ffurano** [comment]
I checked, actually the library already does it, for every new client session, so this attempt to do some additional reinitialization in the code failed, as I would not know which additional openssl reinitialization to look for

**Comment by olifre** [comment]
Hi @ffurano,
many thanks for the investigations!
This is indeed a strange issue - since we now employ the workaround, we did not see it again. I think we need to have and parse the CRLs for the Grid use case, though, so leaving them out is not really an option. 
We will then keep the workaround in place (it does not really harm, and we don't have a good idea on how to track down the actual cause). Thanks!

**Comment by ffurano** [comment]
Well, if some openssl guru has an idea about what to look for, I'll be happy to do it...

**Comment by abh3** [comment]
Hmmm, I think we can close this can we not? The problem is understood and there will be no fix forthcoming.

**Comment by esindril** [comment]
This should not be closed. This should be fixed so that the same behavior is observed when using either xrootd or http. At the moment one would need to regularly restart the xrootd daemon to have this working properly and for a service like EOS, for example, this would mean restarting the manager node regularly which is not an option. 

Also the difference in behavior is confusing since any access over the xrootd protocol works fine but doing the same thing, with the same certificate over HTTP fails with "certificate expired" error.

The HTTP part should probably make use of the same mechanisms that XrdSecgsi uses to avoid such issues with expired CRLs. Given the upcoming R5 I think this is the perfect timing to properly address this ... since I assume 4.11.* branch will never see this fixed. Thanks!

**Comment by abh3** [comment]
OK, this has been kicking around for a while. Could some one investigate and provide guidance on how we can refresh the CRL's? Looking at the code it doesn't seem there is an explicit way of doing this. I haven't even found any information on CRL refreshing on the web for any OpenSSL based product. GSI does it but it's a brute force solution. No small wonder as I can't see any OpenSSL API to do it otherwise. I'd like to get this wrapped up and in R5 as it's a constant pain but I need more suggestions than complaints.

**Comment by olifre** [comment]
@abh3 : While I am also not an OpenSSL expert, looking at what programmers more expert in SSL than me are doing (hostapd, who should be optimizing also for embedded machines):
https://www.spinics.net/lists/hostap/msg04819.html
It seems that flushing the store and reloading is the only way.

**Comment by bbockelm** [comment]
I wrote some tips to the mail list a few days back; copied here for reference:

> ...
> 
> Particularly, the singleton SSL_CTX* here:
> 
> https://github.com/xrootd/xrootd/blob/master/src/XrdHttp/XrdHttpProtocol.cc#L1659
> 
> holds a pointer to a X509_STORE object.  This object loads the CRLs at initialization but never reloads them.  The SSL_CTX* is passed to libvomsapi through the callback registered by libXrdHTTPVOMS; in a finite amount of time, the CRLs expire.
> 
> An example of loading a new X509_STORE is here:
> 
> https://github.com/opensciencegrid/xrootd-lcmaps/blob/master/src/GlobusSupport.cc#L295
> 
> Unfortunately, to be able to know when you can safely delete the old object, you either need a stop-the-world global mutex wrapping the object or you can hash a few across threads as in here:
> 
> https://github.com/opensciencegrid/xrootd-lcmaps/blob/master/src/GlobusSupport.cc#L354
> 
> In that example, we only lock across 1/63 of the threads when we are reloading.

Brian

**Comment by abh3** [comment]
Hi Brian,

Ah yes, I lost that email. So, is it the case that creating  an ssl 
context also loads teh CRL's for that context. Thus, any time I create a 
new context I will get a fresh set of CRL's for that context. Anyway, let 
me inspect the code you pointed to. It will be part of the super merge 
into vomsxrd to have one thing to support.

Andy

**Comment by abh3** [comment]
Hi Oliver,

Thans for the pointer. OK, I have enough to go on I think to nail this.

Andy

**Comment by abh3** [comment]
OK, I will give a go on this. Here is the idea based on hostAPD (thank you Oliver for that crucial pointer) which we can freely use as it's under a BSD license as opposed to other software packages. It also provides an idea on how to simplify the whole refresh process.

From my understanding, the SSL_CTX object is associated with one X509_store that holds the CA certs and CRLS. Typically, there is one such object and it is used with all new connections. During connection processing the information in the X509_store is used to verify the incoming cert. Because, there is one store, if you want to refresh that store you need to lock everything down until that happens; a rather brutal affair. This is why LCMAPS decided to have 64 contexts so only 1/64 of possible connections are blocked when the store has to get refreshed. I think this overly complex and based on what I see in hostAPD there is a much easier way of doing this.

Since in XRootD we have centralized all the TLS handling; so, we have the ability to maintain a shadow context. Everyone who needs to setup a connection asks the central context object for the SSL_CTX. We can hand that out and maintain a reference count to it. When the connection is closed, that context has to be returned and we decrease the reference count. So, we know when it can be deleted. If we need to refresh the CRL’s we do so on the shadow SSL_CTX. Once we set that up the shadow SSL_CTX replaces the one we will be giving out and we decrease the reference count to the old one as we no longer need it. It will get deleted when all uses of it are gone. This allows us to always give out an SSL_CTX object with a current X509_store. 

So, essentially we are doing rolling refreshes and always have an SSL_CTX object that is current for new incoming connections. That way we don’t need to block any threads due to a CRL refresh. Does all of this sound reasonable to everyone?

This is rather trivial to do in R5 but in R4 we don’t have that framework. I think I can do a minimal implementation that would work for the R4 series so I don’t waste time doing it all twice over. Mind you, the only person at the moment that needs “fresh” CRL’s is HTTPS because it uses the presented cert for authentication. XRootD uses GSI which already refreshes CRL’s on its own.

**Comment by olifre** [comment]
@abh3 That sounds very reasonable to me, it also sounds like a very "lightweight" solution in terms of CPU power and without additional heavy locking. Well thought out. 

I am not sure how high the risk is that the reference count never / rarely ever goes to zero (on very busy servers with long-lived connections), that would be the only potential issue I could think about. If that is a real issue, it might be good to echo a warning into the logs if a refresh is long overdue and cannot be made due to the reference count not dropping. 

The proposed timescale is also perfectly fine with me (minimal in R4, good solution in R5), but @esindril is probably the more heavy affected one here, so I'll leave that judgement to him.

**Comment by abh3** [comment]
@olifre I don't think that's really is needed. The SSL_CTX is always current when the connection is made. The CRLs only need to be current when you extract the certificate and try to authenticate it. That only happens once when you make the connection. After that, it doesn't matter if the CRLs go stale because you'll never use them (OK, you may need to when renegotiating but we don't care because the authentication has already been done). So, every connection is guaranteed to have fresh CRLs. The only downside is that you may wind up with multiple SSL_CTX objects as the CRL refresh generates a new SSL_CTX for future connections. Generally long running connections are not that long running (typically less than a day) and CRLs likely don't change often enough to warrant a very short refresh interval (I wish we could determine if CRLs actually changed but we can't).  So, you just wind up sitting on some extra memory for a while. If that creates a problem we can simply add a method to the framework that notifies it that the SSL_CTX use for authentication finished. If that SSL_CTX is no longer current and we have no outstanding authentications we delete the X509_store as it's the biggest memory hog in OpenSSL. Frankly, if the system is very active there will be a high reuse of SSL_CTX objects and if the system is relatively idle it won't create a stress. Of course, somewhere in between lies the big unknown.

**Comment by bbockelm** [comment]
I think I spot an alternate option.

It looks like `SSL_set1_verify_cert_store` overrides in the `SSL` object the `X509_STORE` provided by the global `SSL_CTX` object.  Further, OpenSSL does an atomic refcount on the `X509_STORE` copies.

So, one scheme could be:

1.  Have a global `X509_STORE *current_x509_store` object.  Each time a new `SSL` object is made, grab a read lock, invoke `SSL_set1_verify_cert_store(ssl, current_x509_store)`, and then release the read lock.
2.  Have a 'refresh' thread that, once an hour, creates a new `X509_STORE *new_x509_store = X509_STORE_new()`, grabs a write lock, calls `X509_STORE_free(current_x509_store)`, does `current_x509_store = new_x509_store`, then release the write lock. 

The refcounts are then handled by OpenSSL as opposed to managing them from the Xrootd side; the only blocking is, once an hour, decrementing a refcount and doing an atomic update of a pointer location.  Looking at the glibc implementation of the read-write, it seems the cost of the 'read lock' is an atomic increment except in the case when there's a pending write (which again, happens once an hour).

**Comment by bbockelm** [comment]
@abh3 - https://github.com/xrootd/xrootd/pull/1183 is a minimally-tested version of the above.

The nice part of this approach is it backports cleanly to 4.x.

**Comment by abh3** [comment]
@bbockelm Well, yes, I could do that. However, I would opt for a slightly (though a bit more expensive) approach. I was planning on using the framework's Clone() method to create a new context identical to the global one but with refreshed CRLs and CA certs. I could then extract the X509_store from it use it to set the X509 store in "in use" context and then discard the new context. This does two things a) keeps contexts and connections separated, and b) always uses the same code path to build the new X509_store. So, this keeps it maintainable and dealing with OpenSSL anything that makes it more maintainable is a huge plus. I don't know if you looked at the R5 TLS framework but it completely abstracts out the notion that you are using OpenSSL. We decided to go that route because of ever changing OpenSSL API's and we didn't want to splatter that stuff all over the code. While some of OpenSSL will need to leak into the R4 back port it will be very clean in R5.

**Comment by abh3** [comment]
@bbockelm One more thing. It's just a matter of time that a host certificate will be stolen and we will be forced to apply timely crls on the client side. So, I have to make sure that we can actually do it in a trivial way (i.e. not changing vast amounts of code). It's all doable and relatively easily portable backwards. As an aside, most browsers don't do so well on that point from what I've read in stackoverflow.

**Comment by abh3** [comment]
@bbockelm Lots of issues all around. Please see my comments on your pull request. This is not as easy as one might think.

**Comment by abh3** [comment]
@olifre @bbockelm Gentleman and all other interested parties. Release deadlines are fast approaching and our options are slowly dwindling. The one thing that is becoming very clear is that providing crl refresh for older OpenSSL versions is going to be more difficult that one would have hoped. The upshot is that we can easily provide refresh for RH7 and above but not for vanilla RH6. The argument could be made that this is OK since RH6 sites should have installed more recent versions of OpenSSL due to numerous security issues present in the default install. Furthermore, these sites are few and they never had crl refresh before so perhaps they can continue without. Who knows, maybe all their servers are running RH7 and it's only their their batch farms, which make the matter moot.

I am sympathetic to both of these arguments. I would rather not hold up this feature due to a few sites that already have severe security issues anyway. Well, that's my feeling at the moment. I would like all of your opinions whether to take this approach or not. If we do, we could have something in short order, perhaps even for 4.12.0 with a possible back port. Otherwise, it looks like at least a 3 week delay.

**Comment by esindril** [comment]
As far as we're concerned on the EOS side - RH6/SLC6 support is not required.

**Comment by olifre** [comment]
@abh3 From my point of view, supporting only >/RH7 is fine. I'm not sure if there really are sites who run RH6 on their transfer nodes / file servers still. Of course, it might be that they have just installed them together with the batch nodes, and never upgraded (it happens...). But we can also argue CentOS6 is officially out of support in half a year anyways, and indeed for machines exposed to the internet, it's already now not a good diea to run it. 

On our end (but we are only a single Tier3, so it's hard to extrapolate from my point of view), we are now using containers for any batch jobs and stay up to date on newest RH7 releases. RH8 is still a no-go for us (mainly, since we need CephFS which only supports this in the very latest versions and rely on the WLCG repositories for VOMS setup who are not ready yet). However, once those issues are sorted out, we want to upgrade as soon as possible. I think many sites in high energy physics are trying to decouple OS upgrades from the experiments' software stack using containers now, so this is a growing trend. 

So in short, my opinion is to go >= RH7 only, since it would eat a lot of developer energies to still support the dead RH6 platform. Restarting via a cronjob is still an option if somebody really runs RH6 until it goes EoL anyways. It could be interpreted as a "deprecation warning" ;-).

**Comment by olifre** [comment]
We've upgraded to 4.12.1 now and disabled the daily restart cronjob, so we'll learn in a few days if:
https://github.com/xrootd/xrootd/pull/1183
was effective. Thanks @bbockelm for the change!

**Comment by olifre** [comment]
It really seems the PR was effective! Thanks again to @bbockelm , closing this. 
@esindril , if it did not help for you, of course just reopen.
